### PR TITLE
Implement parsing for records

### PR DESCRIPTION
**Pull request by agocke**
I recommend reviewing the first commit separately. All it does is split DeclarationModifiers into its own field in our flags structs since we ran out of space, and splitting the modifiers out was simpler and clearer. The second commit actually implements the parsing.

As agreed in LDM, the specification requires that a data contextual
keyword be present in front of the class and a parameter list to be
present after the identifier, for the type to be a record.

The absence of either of these two elements will be a semantic error,
not a parsing error.

Relates to https://github.com/dotnet/roslyn/issues/40726 (test plan for records)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 1]
```diff
@@ -24,63 +24,50 @@ internal abstract partial class SourceMemberContainerTypeSymbol : NamedTypeSymbo
         // The flags type is used to compact many different bits of information efficiently.
         private struct Flags
         {
-            // We current pack everything into two 32-bit ints; layouts for each are given below.
-
-            // First int:
+            // We current pack everything into one 32-bit int; layout is given below.
             //
-            // | |d|yy|xxxxxxxxxxxxxxxxxxxxxxx|wwwwww|
+            // |               |vvv|zzzz|f|d|yy||wwwwww|
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: >| [](start = 48, length = 1)

Extra vertical bar

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -1485,6 +1527,8 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
                             semicolon);
 
                     case SyntaxKind.InterfaceKeyword:
+                        RoslynDebug.Assert(openBrace != null);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >RoslynDebug [](start = 24, length = 11)

I don't see how this assertion is correct.  It appears that the above code would accept `interface A;`  I think we either need to make the curly braces optional or make missing tokens for them with an error.

Comment (gafter) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Similarly this code appears to be dropping the parameter list on the floor.  I think it needs to be either added to the syntax, not permitted by the parser, or you need to build a class declaration when that happens.

---
In reply to: [361761780](https://github.com/dotnet/roslyn/pull/40467#discussion_r361761780) [](ancestors = 361761780)

File: src/Compilers/CSharp/Portable/Syntax/ClassDeclarationSyntax.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            => SyntaxFactory.ClassDeclaration(
+                attributeLists,
+                modifiers,
+                SyntaxFactory.Token(SyntaxKind.ClassKeyword),
+                identifier,
+                typeParameterList,
+                parameterList,
+                baseList,
+                constraintClauses,
+                openBraceToken: default,
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/ClassDeclarationSyntax.cs]: >default [](start = 32, length = 7)

This doesn't look right.  It creates a node without braces nor semicolon.

File: src/Compilers/CSharp/Portable/Syntax/StructDeclarationSyntax.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                SyntaxFactory.Token(SyntaxKind.StructKeyword),
+                identifier,
+                typeParameterList,
+                parameterList,
+                baseList,
+                constraintClauses,
+                default,
+                members,
+                default,
+                default);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/StructDeclarationSyntax.cs]: >default [](start = 16, length = 7)

I suspect you intended this to be the creation of a fresh semicolon token.

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs [hunk 5]
```diff
@@ -156,7 +156,7 @@ public void DeeplyNestedGeneric()
                 (ExecutionArchitecture.x86, ExecutionConfiguration.Debug) => 270,
                 (ExecutionArchitecture.x86, ExecutionConfiguration.Release) => 1290,
                 (ExecutionArchitecture.x64, ExecutionConfiguration.Debug) => 170,
-                (ExecutionArchitecture.x64, ExecutionConfiguration.Release) => 730,
+                (ExecutionArchitecture.x64, ExecutionConfiguration.Release) => 700, // PROTOTYPE(angocke): Improve stack usage
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: >// PROTOTYPE(angocke): Improve stack usage [](start = 84, length = 42)

I wouldn't bother.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
+{
+    [CompilerTrait(CompilerFeature.ReadOnlyReferences)]
+    public class RecordTests : CompilingTestBase
+    {
+        [Fact]
+        public void RecordLanguageVersion()
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs]: >RecordLanguageVersion [](start = 20, length = 21)

Also try `class Point;`.  That should not be permitted in older language versions either.

**Comment by gafter** [comment]
Do you have a document that describes the feature this is intended to implement?  Neither [https://github.com/dotnet/csharplang/blob/master/proposals/records.md](https://github.com/dotnet/csharplang/blob/master/proposals/records.md) nor [https://github.com/dotnet/csharplang/blob/master/proposals/recordsv2.md](https://github.com/dotnet/csharplang/blob/master/proposals/recordsv2.md) seems to be right.

**Comment by gafter** [comment]
One of the things I'd be looking for in the specification is how you will modify the syntax and parser to handle constructor parameters of the base type. Without knowing how that's going to be handled, I cannot know if these parser changes implement the intended design.

---
In reply to: [569365350](https://github.com/dotnet/roslyn/pull/40467#issuecomment-569365350) [](ancestors = 569365350)

**Comment by gafter** [comment]
Finished review pass (Iteration 2).

File: eng/targets/Settings.props [hunk 7]
```diff
@@ -17,7 +17,7 @@
     <DisableImplicitNuGetFallbackFolder>true</DisableImplicitNuGetFallbackFolder>
     <ToolsetPackagesDir>$(RepoRoot)build\ToolsetPackages\</ToolsetPackagesDir>
 
-    <RoslynPortableTargetFrameworks>net472;netcoreapp2.1</RoslynPortableTargetFrameworks>
+    <RoslynPortableTargetFrameworks>netcoreapp2.1;net472</RoslynPortableTargetFrameworks>
```
Comment (333fred) [inline eng/targets/Settings.props]: Please revert this change. Quoting from @jaredpar a few months ago:
> We need to keep net472 as the first TF listed because that controls items like which tests run by default.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -1516,6 +1560,8 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
             }
         }
 
+#nullable disable
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: `restore`

**Comment by 333fred** [comment]
Done review pass (commit 3)

File: src/Compilers/CSharp/Portable/Syntax/ClassDeclarationSyntax.cs (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Syntax/ClassDeclarationSyntax.cs]: However, since you don't know the language version at this point, perhaps this should create brace tokens always.

---
In reply to: [361762456](https://github.com/dotnet/roslyn/pull/40467#discussion_r361762456) [](ancestors = 361762456)

File: eng/targets/Settings.props (hunk 7 above)
Comment (agocke) [reply eng/targets/Settings.props]: Yeah, that's the point. We should swap the default. I can't test on Mac using VS Code unless it's first.

Comment (jaredpar) [reply eng/targets/Settings.props]: > We should swap the default. I can't test on Mac using VS Code unless it's first.

Definitely don't want us to do this. As @333fred noted, the order here is deliberate. Changing it will not fix anything, instead it will just fix one scenario and break another. In this case it would be fixing Mac, the minority scenario, and breaking Windows, the dominant scenario. 

Where is the bug on VS Code that prevents Mac testing here? Have you filed a bug?

Comment (agocke) [reply eng/targets/Settings.props]: Why would it break Windows? The test explorer seems to work fine

Comment (jaredpar) [reply eng/targets/Settings.props]: > Why would it break Windows?

The set of tests run for desktop and core are significantly different. Moving to core by default will break the expectations of developers that expect to get the desktop tests.

If you think the trade off is a good one here then the correct course of action is to discuss the policy with the team. Not change it subtly in a PR. Particularly since this has come up several times before and we've kept the current order for these reasons.

> Doesn't look like there's a bug here for VS Code. 

VS Code cannot test code that every other tool in our toolbox can test. Is the explicit design of VS Code that they can't support multi-targeted projects? If so that seems really unfortunate. 

> it's a feature they don't have.

Where is the issue tracking the feature request?

Comment (agocke) [reply eng/targets/Settings.props]: > If you think the trade off is a good one here then the correct course of action is to discuss the policy with the team. Not change it subtly in a PR. Particularly since this has come up several times before and we've kept the current order for these reasons.

Yup, this is just going into a feature branch. I had planned to send an email out before changing it in master. Can I leave a PROTOTYPE comment above and we can do that when I'm back in Seattle?

> Where is the issue tracking the feature request?

One thing we have to change for that is running our desktop tests as 64-bit by default. For the life of me I cannot figure out how to change that from 32-bit. I gave up after a couple hours.

https://github.com/OmniSharp/omnisharp-vscode/issues/1716 

> You aren't doing anything wrong. x86 debugging is not supported by the C# extension. You will need full Visual Studio for that.

Beyond that, I'm not sure what feature we would be asking for. Some way to change the framework the tests are run under? VS only got that a couple months ago, so it's not exactly something I'd put at the top of the VS Code priority list.

Comment (jaredpar) [reply eng/targets/Settings.props]: > Can I leave a PROTOTYPE comment above and we can do that when I'm back in Seattle?

That's fine by me. 

> Beyond that, I'm not sure what feature we would be asking for. Some way to change the framework the tests are run under? 

I think the ask is pretty straight forward: we have a project which is multi-targeted to desktop and .NET Core. When that happens don't fail, instead prefer the .NET Core target framework. Basically do the only action which is capable of succeeding in this environment.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 9]
```diff
@@ -1391,69 +1422,80 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
...
                 }
 
-                var openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
-
-                // ignore members if missing type name or missing open curly
-                if (name.IsMissing || openBrace.IsMissing)
+                SyntaxToken semicolon;
+                SyntaxToken? openBrace;
+                SyntaxToken? closeBrace;
+                if (isInterface || CurrentToken.Kind != SyntaxKind.SemicolonToken)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >CurrentToken.Kind != SyntaxKind.SemicolonToken [](start = 35, length = 46)

I feel like this isn't a great solution. I know this change was prompted by Neal, but I feel like a better solution would be to parse the interface as a record declaration, and then issue a semantic error for "Interfaces cannot be records."

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                // (2,31): error CS8652: The feature 'records' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                // data class Point(int x, int y);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, ";").WithArguments("records").WithLocation(2, 31)
+            );
+
+            comp = CreateCompilation(src1, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            comp = CreateCompilation(src2, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            comp = CreateCompilation(src3, parseOptions: TestOptions.RegularPreview);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs]: Missing a `VerifyDiagnostics`

**Comment by 333fred** [comment]
Done review pass (commit 5)

**Comment by gafter** [comment]
Similarly, I would expect to see a syntax tree and parsing for the "constructor body" from [https://github.com/dotnet/csharplang/blob/master/proposals/records.md#primary-constructor-body](https://github.com/dotnet/csharplang/blob/master/proposals/records.md#primary-constructor-body) (which incidentally was copied verbatim into Java's record feature).

---
In reply to: [569365766](https://github.com/dotnet/roslyn/pull/40467#issuecomment-569365766) [](ancestors = 569365766,569365350)

**Comment by agocke** [comment]
@gafter Good point, I've added https://github.com/dotnet/csharplang/blob/master/proposals/records-wip.md

Basically, I intend to support the syntax that you linked, with the added `data` modifier, but I'm trying to carve out a few minimal work items to avoid making giant PRs.
